observer/observer.go

```diff
@@ -152,14 +158,56 @@ func (o *Observer) Watch(files []string) error {
 	return nil
 }
 
-// handleEvent handle an event.
-func (o *Observer) handleEvent(event interface{}) {
-	// Run all listeners for this event.
+// SetBufferDuration set the buffer damping duration
+func (o *Observer) SetBufferDuration(d time.Duration) {
+	// Create the buffer mutex, if missing.
+	if o.bufferMutex == nil {
+		o.bufferMutex = &sync.Mutex{}
+	}
+
+	// Set the buffer duration.
+	o.bufferDuration = d
+}
+
+// sendEvent send one or more events to the observer listeners.
+func (o *Observer) sendEvent(event interface{}) {
 	for _, listener := range o.listeners {
 		go listener(event)
 	}
 }
 
+// handleEvent handle an event.
+func (o *Observer) handleEvent(event interface{}) {
+	// If we do not buffer events, just send this event now.
+	if o.bufferDuration == 0 {
+		o.sendEvent(event)
+		return
+	}
+
+	// Lock this function.
+	o.bufferMutex.Lock()
+	defer o.bufferMutex.Unlock()
+
+	// Add new event to the event buffer
+	o.bufferEvents = append(o.bufferEvents, event)
+
+	// If this is the first event, set a timeout function.
+	if o.bufferTimer == nil {
+		o.bufferTimer = time.AfterFunc(o.bufferDuration, func() {
+			// Lock this function.
+			o.bufferMutex.Lock()
+			defer o.bufferMutex.Unlock()
+
+			// Send all events in event buffer
+			o.sendEvent(o.bufferEvents)
+
+			// Reset events buffer
+			o.bufferTimer = nil
```

Comment:
The current implementation is how I would do it in callbacks.

But here I would create a "consume" like buffer, that only after I know for sure that something consume it I'll remove the buffer.

The reason is that goroutines are actually green threads, and you can enter to one routine but not execute the actual code yet, and you might already clearing the content.

If you can, just add a channel to know it was consume, or "routing ended" to clear it up.

Reply:
Thanks ! this line was removed in current code, things changed since this PR was merged :-)